pyo3-build-config/src/impl_.rs

```diff
@@ -486,47 +549,58 @@ pub fn any_cross_compiling_env_vars_set() -> bool {
         || env::var_os("PYO3_CROSS_PYTHON_VERSION").is_some()
 }
 
-fn cross_compiling() -> Result<Option<CrossCompileConfig>> {
+fn cross_compiling_from_cargo_env() -> Result<Option<CrossCompileConfig>> {
+    let host = cargo_env_var("HOST").unwrap();
+    let target = cargo_env_var("TARGET").unwrap();
+
+    if host == target {
+        // definitely not cross compiling
+        return Ok(None);
+    }
+
+    let target_arch = cargo_env_var("CARGO_CFG_TARGET_ARCH").unwrap();
+    let target_vendor = cargo_env_var("CARGO_CFG_TARGET_VENDOR").unwrap();
+    let target_os = cargo_env_var("CARGO_CFG_TARGET_OS").unwrap();
+
+    cross_compiling(&host, &target_vendor, &target_arch, &target_os)
+}
+
+/// Detect whether we are cross compiling and return an assembled CrossCompileConfig if so.
+pub fn cross_compiling(
+    host: &str,
+    target_arch: &str,
+    target_vendor: &str,
+    target_os: &str,
+) -> Result<Option<CrossCompileConfig>> {
     let cross = env_var("PYO3_CROSS");
     let cross_lib_dir = env_var("PYO3_CROSS_LIB_DIR");
     let cross_python_version = env_var("PYO3_CROSS_PYTHON_VERSION");
 
-    let target_arch = cargo_env_var("CARGO_CFG_TARGET_ARCH");
-    let target_vendor = cargo_env_var("CARGO_CFG_TARGET_VENDOR");
-    let target_os = cargo_env_var("CARGO_CFG_TARGET_OS");
+    let target_triple = format!("{}-{}-{}", target_arch, target_vendor, target_os);
 
     if cross.is_none() && cross_lib_dir.is_none() && cross_python_version.is_none() {
         // No cross-compiling environment variables set; try to determine if this is a known case
         // which is not cross-compilation.
 
-        let target = cargo_env_var("TARGET").unwrap();
-        let host = cargo_env_var("HOST").unwrap();
-        if target == host {
-            // Not cross-compiling
-            return Ok(None);
-        }
-
-        if target == "i686-pc-windows-msvc" && host == "x86_64-pc-windows-msvc" {
+        if target_triple == "i686-pc-windows-msvc" && host == "x86_64-pc-windows-msvc" {
```

Comment:
For now I think this check could instead use `starts_with` and a decision on using `target_lexicon` can be put off.

I understand the hesitation on adding a dependency, especially without a concrete example of new functionality it would add.

Reply:
Hm, that still misses the `msvc` part though which is probably important for this check. That specific check could also move back to the `cross_compiling_from_cargo_env` function.  I'll give this a bit more thought.